### ISSUE TITLE
v2 Kata agent allocating a tty even though container is non-interactive

### ISSUE THREAD

**Issue by skaegi** [kata-containers/kata-containers#3518]
/dev/tty is allocated and live with the v2 agent even when using a standard non-interactive container. This is different behavior than the v1 agent as well as regular runc. This was confirmed in all 2.2 and 2.3.1 agents with our internal builds which are using alpine as the guest os however I believe the problem is not specific to our build.

--

We have a customer using [sbt](https://www.scala-sbt.org/) for a scala build. Under the covers this tools can operate interactively and does a check to see if /dev/tty is "live", and if so tries to configure it. Since a kube process is actually a background process this ends up resulting in a SIGTTOU which suspends the process sequence ultimately hanging the build process. More info [here](http://curiousthing.org/sigttin-sigttou-deep-dive-linux) 

Our parent process should be fixed to handle this signal similar to work done for tini however this worked correctly with the v1 agent and is currently breaking our customer's use.

Here's a minimal pod to illustrate...
```
apiVersion: v1
kind: Pod
metadata:
  name: untrusted-sbt-min
spec:
  containers:
    - name: sbt
      image: hseeberger/scala-sbt:8u265_1.3.13_2.12.12
      imagePullPolicy: Always
      command: ["sh", "-c"]
      args:
        - ls -al /dev/tty && sh -c "stty -g < /dev/tty"
      #stdin: true
      #tty: true
  nodeSelector:
    katacontainers.io/kata-runtime: "true"
  restartPolicy: Never
  runtimeClassName: kata-qemu
  ```
  
  When run this logs...
  ```
  $ k logs untrusted-sbt-min 
crw-rw-rw- 1 root root 5, 0 Jan 20 20:59 /dev/tty
500:5:4bf:8a3b:3:1c:7f:15:4:0:1:0:11:13:1a:0:12:f:17:16:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0
```
... which is incorrect as the tty should not be allocated.

 If you remove the `runtimeClassName` or use a v1 agent you will get...
```
$ k logs untrusted-sbt-min 
crw-rw-rw- 1 root root 5, 0 Jan 20 20:59 /dev/tty
sh: 1: cannot open /dev/tty: No such device or address
```
... which is correct as the tty should not be allocated and live.

If you now uncomment the `stdin: true` and `tty: true` while using runc or a v1 agent and re-run you will get...
```
$ k logs untrusted-sbt-min 
crw-rw-rw- 1 root root 5, 0 Jan 20 21:09 /dev/tty
500:5:bf:8a3b:3:1c:7f:15:4:0:1:0:11:13:1a:0:12:f:17:16:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0
```
... which makes sense as we do want the tty allocated.

**Comment by fidencio** [comment]
It seems that the agent is purposefully opening `/dev/tty`, see: https://github.com/kata-containers/kata-containers/blob/618aa659d652a18f1fcd9b43f5dc9bdc15d6904b/src/agent/rustjail/src/container.rs#L163-L171 and it seems to be like that since f8ced638d253bbae63d2cdfc35eb873dfb547e59.

/cc @lifupan @liubin, the patch to be proposed is a quite simple one, but what's the intention for having `/dev/tty` opened by default?

**Comment by skaegi** [comment]
Hmm... if there is a simple patch for this I'd be happy to rebuild and try it.

 I looked briefly at the code and I think the creation of the device is what I'd expect... the connecting up also looked correct here... https://github.com/kata-containers/kata-containers/blob/618aa659d652a18f1fcd9b43f5dc9bdc15d6904b/src/agent/rustjail/src/container.rs#L675-L679
 ...  unless something like the deserialization was busted.

@lifupan I'm happy to write the tests here. Let me know if there is there anything I could to do to help.

**Comment by lifupan** [comment]
Hi @skaegi 

It's weird, I had tried with kata-2.3.0 and I couldn't reproduce this issue. I used the following pod yaml:

lifupan@kant-server:~$ cat pod.yaml 
```
apiVersion: v1
kind: Pod
metadata:
  name: untrusted-sbt-min
spec:
  containers:
    - name: sbt
      image: ubuntu:latest
      imagePullPolicy: Always
      command: ["sh", "-c"]
      args:
        - ls -al /dev/tty && sh -c "stty -g < /dev/tty"
          #stdin: true
          #tty: true
  restartPolicy: Never
  runtimeClassName: kata
```

and I actually got the following logs with: 
lifupan@kant-server:~$ sudo kubectl logs untrusted-sbt-min 

```
crw-rw-rw- 1 root root 5, 0 Jan 25 12:23 /dev/tty
sh: 1: cannot open /dev/tty: No such device or address
```

**Comment by skaegi** [comment]
hmm... ok thanks @lifupan -- let me do some more debugging on my side and try the release 2.3.0 release image etc. Perhaps my guest os (alpine) is somehow triggering the problem?? I'll try changing that too.

**Comment by skaegi** [comment]
So... what I found is that:
1) the 2.3.0 release alpine rd_image produced exactly the same incorrect result as I saw
2) the 2.3.0 clear linux image produced the correct result (also validated with tty=true and tty=false)
3) the 2.3.0 clear linux image configured with `kernel_params = "init=/usr/bin/kata-agent"` again produced the incorrect tty result I saw

I'm really not sure why but it seems like when the agent is launched by `systemd` things are working correctly but when the agent is launched as the init process things do not work correctly.

**Comment by skaegi** [comment]
Reading and thinking I suspect one possibility for what's happening for the agent "init" case is that `/init` is given a controlling terminal and the containers launched by the agent are "unexpectedly" inheriting the controlling terminal from the parent. Systemd and essentially all similar daemon launchers protect against that and do not give the child "service" processes a controlling terminal. This is just a theory of course as I'm just beginning to understand the code ;)

**Comment by lifupan** [comment]
Hi @skaegi 

I had tried to investigate this issue, and I can confirm the latest kata-agent had fixed this issue, and you must meet it using the kata-agent before merged this issue PR : https://github.com/kata-containers/kata-containers/pull/3064 .

And I had tried to reproduce it with revert https://github.com/kata-containers/kata-containers/pull/3064, and I did can reproduce your issue, but when I used the kata-agent applied https://github.com/kata-containers/kata-containers/pull/3064, both of the  alpine initrd and clear linux image configured with kernel_params = "init=/usr/bin/kata-agent" couldn't reproduce your issue.

**Comment by skaegi** [comment]
Thanks so much @lifupan -- confirmed that with 2.3.1 after applying the fix in #3064 everything worked correctly. This issue I think can be closed as it's fixed already on `main`. I'll add a note to #3064 asking for a back-port to 2.3.x.
